zcash_client_backend/src/data_api.rs

```diff
@@ -61,6 +65,30 @@ pub trait WalletRead {
     /// This will return `Ok(None)` if no block data is present in the database.
     fn block_height_extrema(&self) -> Result<Option<(BlockHeight, BlockHeight)>, Self::Error>;
 
+    /// Returns the height to which the wallet has been fully scanned.
+    ///
+    /// This is the height for which the wallet has fully trial-decrypted this and all preceding
+    /// blocks above the wallet's birthday height. Along with this height, this method returns
+    /// metadata describing the state of the wallet's note commitment trees as of the end of that
+    /// block.
+    fn fully_scanned_height(
+        &self,
+    ) -> Result<Option<(BlockHeight, chain::CommitmentTreeMeta)>, Self::Error>;
+
+    /// Returns a vector of suggested scan ranges based upon the current wallet state.
+    ///
+    /// This method should only be used in cases where the [`CompactBlock`] data that will be made
+    /// available to `scan_cached_blocks` for the requested block ranges includes note commitment
+    /// tree size information for each block; or else the scan is likely to fail if notes belonging
+    /// to the wallet are detected.
+    ///
+    /// [`CompactBlock`]: crate::proto::compact_formats::CompactBlock
+    fn suggest_scan_ranges(
+        &self,
+        batch_size: usize,
+        limit: usize,
```

Comment:
I think a `limit` parameter that sets a maximum number of ranges to be returned _might_ be useful in highly fragmented wallet note commitment tree situations, but it's kind of meaningless without having control over the batch size. Alternately the limit could be a "soft" limit on the number of blocks represented by the ranges being returned, whereby the returned ranges might exceed the requested number of blocks, but this is also not very useful; it would be a valid implementation of this API to simply return a single range representing an incomplete fully-linear scan over chain history. So I guess that it makes the most sense to take no arguments at all to this function for now.

Reply:
In any case, if these arguments are present then they should be documented.